import required libraries.

In [39]:
# import numpy as np
import pandas as pd
from sklearn import preprocessing


Read data

In [2]:
data=pd.read_csv('dataset.csv')

In [3]:
# display data head
data.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

In [4]:
# Gathering every feature's data type abd other informations
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           20640 non-null  float64
 1   latitude            20640 non-null  float64
 2   housing_median_age  20640 non-null  float64
 3   total_rooms         20640 non-null  float64
 4   total_bedrooms      20433 non-null  float64
 5   population          20640 non-null  float64
 6   households          20640 non-null  float64
 7   median_income       20640 non-null  float64
 8   median_house_value  20640 non-null  float64
 9   ocean_proximity     20640 non-null  object 
dtypes: float64(9), object(1)
memory usage: 1.6+ MB


In [6]:
data.describe().T

count           mean            std         min  \
longitude           20640.0    -119.569704       2.003532   -124.3500   
latitude            20640.0      35.631861       2.135952     32.5400   
housing_median_age  20640.0      28.639486      12.585558      1.0000   
total_rooms         20640.0    2635.763081    2181.615252      2.0000   
total_bedrooms      20433.0     537.870553     421.385070      1.0000   
population          20640.0    1425.476744    1132.462122      3.0000   
households          20640.0     499.539680     382.329753      1.0000   
median_income       20640.0       3.870671       1.899822      0.4999   
median_house_value  20640.0  206855.816909  115395.615874  14999.0000   

                            25%          50%           75%          max  
longitude             -121.8000    -118.4900    -118.01000    -114.3100  
latitude                33.9300      34.2600      37.71000      41.9500  
housing_median_age      18.0000      29.0000      37.00000      52.0000  
total_rooms           1447.7500    2127.0000    3148.00000   39320.0000  
total_bedrooms         296.0000     435.0000     647.00000    6445.0000  
population             787.0000    1166.0000    1725.00000   35682.0000  
households             280.0000     409.0000     605.00000    6082.0000  
median_income            2.5634       3.5348       4.74325      15.0001  
median_house_value  119600.0000  179700.0000  264725.00000  500001.0000

In [8]:
data.isnull()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0          False     False               False        False           False   
1          False     False               False        False           False   
2          False     False               False        False           False   
3          False     False               False        False           False   
4          False     False               False        False           False   
...          ...       ...                 ...          ...             ...   
20635      False     False               False        False           False   
20636      False     False               False        False           False   
20637      False     False               False        False           False   
20638      False     False               False        False           False   
20639      False     False               False        False           False   

       population  households  median_income  median_house_value  \
0           False       False          False               False   
1           False       False          False               False   
2           False       False          False               False   
3           False       False          False               False   
4           False       False          False               False   
...           ...         ...            ...                 ...   
20635       False       False          False               False   
20636       False       False          False               False   
20637       False       False          False               False   
20638       False       False          False               False   
20639       False       False          False               False   

       ocean_proximity  
0                False  
1                False  
2                False  
3                False  
4                False  
...                ...  
20635            False  
20636            False  
20637            False  
20638            False  
20639            False  

[20640 rows x 10 columns]

In [9]:
data.isnull().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

# Missign value management
Feature total_rooms has missing values. the data type of this feature is float64. 
The total number of missing values is 207.
The total number of entries is 20640.
That means, missing values take around 1% of our data + we have plenty of data, so I don't mind to remove all the rows containing missing value. (removeing that row)
However in this exercise, we replace the missing data with median

In [17]:
# Replacing missing values with median
filled_data = data.fillna(data.median())

C:\Users\Hamidreza\AppData\Local\Temp\ipykernel_10556\3498446321.py:2: FutureWarning: The default value of numeric_only in DataFrame.median is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  filled_data = data.fillna(data.median())


In [18]:
# replacement reasurance
filled_data.isna().sum()

longitude             0
latitude              0
housing_median_age    0
total_rooms           0
total_bedrooms        0
population            0
households            0
median_income         0
median_house_value    0
ocean_proximity       0
dtype: int64

In [19]:
filled_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           20640 non-null  float64
 1   latitude            20640 non-null  float64
 2   housing_median_age  20640 non-null  float64
 3   total_rooms         20640 non-null  float64
 4   total_bedrooms      20640 non-null  float64
 5   population          20640 non-null  float64
 6   households          20640 non-null  float64
 7   median_income       20640 non-null  float64
 8   median_house_value  20640 non-null  float64
 9   ocean_proximity     20640 non-null  object 
dtypes: float64(9), object(1)
memory usage: 1.6+ MB


## Outlier removal

In [49]:
# Select all of the keys that are numerical,thus we assign them to numerical_col
selected_data_columns = filled_data.select_dtypes(np.number)
numeric_col = selected_data_columns.keys()
# selected_data_columns
numeric_col


Index(['longitude', 'latitude', 'housing_median_age', 'total_rooms',
       'total_bedrooms', 'population', 'households', 'median_income',
       'median_house_value'],
      dtype='object')

In [31]:


for x in numeric_col:
    # q75,q25 = np.percentile(heart.loc[:,x],[75,25])  # a box plot of the quartile range and min/max values method
    # IQR = q75-q25
    # max = q75+(1.5*IQR)
    # min = q25-(1.5*IQR)

    max_data = filled_data[x].mean() + 3*filled_data[x].std()       # Z-score method
    min_data = filled_data[x].mean() - 3*filled_data[x].std()
 
    filled_data.loc[filled_data[x] < min_data,x] = np.nan   #filling the outliers values with 'nan'
    filled_data.loc[filled_data[x] > max_data,x] = np.nan   #filling the outliers values with 'nan'

In [32]:
filled_data.isna().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms           373
total_bedrooms        375
population            342
households            363
median_income         345
median_house_value      0
ocean_proximity         0
dtype: int64

In [33]:
removed_outliers = filled_data.dropna()

## Standardization (normalization)
Standard Scaler
It standardize features by removing the mean and scaling to unit variance The standard score of a sample x is calculated as:

z = (x - u) / s

First we want to seperate our target column. We consider median_house_value as our target column.<br/>
Second, we seperate our not numerical data.<br/>
Then we use standart scaler formula to standardize our data features.

In [81]:
# remove target
target_data = removed_outliers['median_house_value']
features = removed_outliers.drop(columns='median_house_value')
# remove object data 
objective_features = features['ocean_proximity']
numeric_features = features.drop(columns='ocean_proximity')
numeric_features.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  
0       322.0       126.0         8.3252  
1      2401.0      1138.0         8.3014  
2       496.0       177.0         7.2574  
3       558.0       219.0         5.6431  
4       565.0       259.0         3.8462

In [83]:
# Using standard scaler formula
normalized_features = (numeric_features - numeric_features.mean())/numeric_features.std()
normalized_features

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0      -1.320375  1.038968            0.964812    -1.053373       -1.258591   
1      -1.315382  1.029643           -0.645900     3.281869        2.118273   
2      -1.325367  1.024980            1.850704    -0.644178       -1.047753   
3      -1.330360  1.024980            1.850704    -0.778717       -0.892217   
4      -1.330360  1.024980            1.850704    -0.532642       -0.736680   
...          ...       ...                 ...          ...             ...   
20635  -0.751237  1.785020           -0.323758    -0.506153       -0.411782   
20636  -0.811146  1.789682           -0.887507    -1.180942       -1.186007   
20637  -0.816138  1.761705           -0.968042    -0.095563       -0.028126   
20638  -0.866063  1.761705           -0.887507    -0.370219       -0.290810   
20639  -0.826123  1.733729           -1.048578     0.274595        0.424657   

       population  households  median_income  
0       -1.280698   -1.250024       2.871928  
1        1.409283    2.531766       2.857069  
2       -1.055562   -1.059440       2.205302  
3       -0.975341   -0.902488       1.197497  
4       -0.966284   -0.753010       0.075696  
...           ...         ...            ...  
20635   -0.603997   -0.487687      -1.351388  
20636   -1.236706   -1.294867      -0.729274  
20637   -0.394388   -0.102782      -1.264173  
20638   -0.738561   -0.416685      -1.159790  
20639    0.097287    0.259702      -0.834281  

[19746 rows x 8 columns]

In [88]:
normalized_features.describe().T

count          mean  std       min       25%       50%  \
longitude           19746.0 -6.251884e-15  1.0 -2.378772 -1.105700  0.536812   
latitude            19746.0  1.148614e-15  1.0 -1.450978 -0.802846 -0.644310   
housing_median_age  19746.0  1.763223e-17  1.0 -2.256612 -0.806971 -0.001615   
total_rooms         19746.0  6.621081e-17  1.0 -1.665424 -0.669970 -0.215464   
total_bedrooms      19746.0  6.387184e-17  1.0 -1.697549 -0.683971 -0.214770   
population          19746.0 -1.370996e-16  1.0 -1.693446 -0.685512 -0.208069   
households          19746.0 -6.162283e-17  1.0 -1.713405 -0.682008 -0.211153   
median_income       19746.0  1.252248e-16  1.0 -2.013394 -0.738701 -0.145899   

                         75%       max  
longitude           0.781442  2.633636  
latitude            0.969026  2.936736  
housing_median_age  0.642670  1.850704  
total_rooms         0.443118  4.731829  
total_bedrooms      0.448851  4.451317  
population          0.466043  4.536596  
households          0.465234  4.422661  
median_income       0.576508  3.640376

Data normalization transforms multiscaled data to the same scale. After normalization, all variables have a similar influence on the model, improving the stability and performance of the learning algorithm.